### PR TITLE
BuildResidentialHPXML: update apply_defaults and apply_validation

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

We weren't validating _after_ applying defaults.

This PR makes validation updates for the following combinations of `apply_defaults` and `apply_validation`:
```
apply_defaults = false
apply_validation = false
=> no validation

apply_defaults = false
apply_validation = true
=> validation once

apply_defaults = true
apply_validation = false
=> validation once before defaulting

apply_defaults = true
apply_validation = true
=> validation twice before and after defaulting
```

## Checklist

PR Author: Check these when they're done. Not all may apply. ~~strikethrough~~ and check any that do not apply. 

PR Reviewer: Verify each has been completed.

- [ ] ~Schematron validator (`EPvalidator.xml`) has been updated~
- [ ] ~Sample files have been added/updated (via `tasks.rb`)~
- [x] Tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests` and/or `workflow/tests/hpxml_translator_test.rb`)
- [ ] ~Documentation has been updated~
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected changes to simulation results of sample files

### CODE REVIEW

File: BuildResidentialHPXML/measure.rb [hunk 1]
```diff
@@ -3175,10 +3174,6 @@ def run(model, runner, user_arguments)
       return false
     end
 
-    # Write HPXML file again if defaults applied
-    if args[:apply_defaults]
-      XMLHelper.write_file(hpxml_doc, hpxml_path)
-    end
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3181]: 👍

File: BuildResidentialHPXML/measure.rb [hunk 2]
```diff
@@ -3443,17 +3446,23 @@ def self.create(runner, model, args, epw_path, hpxml_path)
...
       HPXMLDefaults.apply(runner, hpxml, eri_version, weather, epw_file: epw_file)
       hpxml_doc = hpxml.to_oga()
+      XMLHelper.write_file(hpxml_doc, hpxml_path)
+    end
+
+    if args[:apply_validation]
+      # Optionally check for invalid HPXML file (with or without defaults applied)
+      if not validate_hpxml(runner, hpxml, hpxml_doc, hpxml_path)
+        return false
+      end
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3465]: 👍 👍

File: BuildResidentialHPXML/measure.rb [hunk 3]
```diff
@@ -3365,6 +3360,14 @@ def argument_errors(args)
     error = (args[:window_aspect_ratio] <= 0)
     errors << 'Window aspect ratio must be greater than zero.' if error
 
+    if (args[:heating_system_type] != 'none' || args[:cooling_system_type] != 'none' || args[:heat_pump_type] != 'none')
+      if (args[:ducts_supply_surface_area].is_initialized && !args[:ducts_return_surface_area].is_initialized) || (args[:ducts_return_surface_area].is_initialized && !args[:ducts_supply_surface_area].is_initialized)
+        errors << 'Must have either both supply/return ducts surface area specified, or neither specified.'
+      elsif (!args[:ducts_supply_surface_area].is_initialized && !args[:ducts_return_surface_area].is_initialized) && (args[:ducts_number_of_return_registers].is_initialized && args[:ducts_number_of_return_registers].get == 0)
+        errors << 'Number of return registers must be greater than zero when both supply/return surface area not specified.'
+      end
+    end
```
[the remark this answers is not in the document]
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb:3369]: The HPXML validation error I was seeing related to this was thrown _after_ defaulting. And validation _after_ defaulting is optional. (ResStock wasn't previously validating after defaulting - should it be?)

I agree we don't _need_ this, and it'd be impossible to write checks up front for all possible errors, but doesn't this convey more information to the user about why they'd see the more cryptic HPXML validation error downstream?

[the remark this answers is not in the document]
Comment (shorowit) [reply BuildResidentialHPXML/measure.rb:3369]: It does, but like I said, you're covering two arbitrary cases you happened to hit out of hundreds of possible other cases. It just doesn't seem worthwhile to me.

**Review by shorowit — APPROVED** [review]

**Pull request merged**